kernel: add __BUG macros for fatal runtime conditions

xodus7 opened the pull request:
There are cases where we to mark conditions as fatal during runtime that do not compile out when asserts are disabled. This change adds the __BUG/__BUG_ON macros that are similar to their Linux kernel counterparts and standardize how these conditions can be marked in the codebase.

xodus7 committed:
kernel: add __BUG macros for fatal runtime conditions

There are cases where we to mark conditions as fatal during runtime
that do not compile out when asserts are disabled. This change adds
the __BUG/__BUG_ON macros that are similar to their Linux kernel
counterparts and standardize how these conditions can be marked in
the codebase.

Signed-off-by: Corey Wharton <xodus7@cwharton.com>

andyross reviewed (commented):
No particular complaints about the API, but I'm really sensitive to "dueling interface" skew.  There should at least be a good faith effort to identify and port existing users of k_panic() to the new API.

Alternatively: maybe augment k_panic instead (make it a macro that logs the location, add a k_panic_on(), etc...)?  I like Linux too, but "BUG" isn't particularly Zephyrese.  Linux has the advantage of always being a conceptually unified tree, where we need to live with apps that might have their own ideas about API schemes.

xodus7 commented:
> No particular complaints about the API, but I'm really sensitive to "dueling interface" skew. There should at least be a good faith effort to identify and port existing users of k_panic() to the new API.
> 
> Alternatively: maybe augment k_panic instead (make it a macro that logs the location, add a k_panic_on(), etc...)? I like Linux too, but "BUG" isn't particularly Zephyrese. Linux has the advantage of always being a conceptually unified tree, where we need to live with apps that might have their own ideas about API schemes.

The intent was not really to replace k_panic but provide a way for fatal conditions to marked that:

1) optionally could still be disabled if desired (CONFIG_BUG_NO_PANIC)
2) optionally add extra logging (CONFIG_BUG_PRINT_LOCATION)
3) Behaves just like ASSERT when those are enabled

k_panic is already a macro so we could satisfy point 2 with a Kconfig option and adding ```k_panic_on``` would also simplify conditionals. But I don't think it would be acceptable to change k_panic for  points 1 or 3.

We could keep things the way they are and let clients fail with printks (if desired) and plain ol' k_panic which is what I did in #80343.

cfriedt commented on include/zephyr/kernel.h:
```diff
@@ -6076,6 +6076,40 @@ static inline void k_cpu_atomic_idle(unsigned int key)
+
+#else
+/* If asserts are enabled, then these behave exactly like asserts. */
+#define __BUG()        __ASSERT_NO_MSG(false)
+#define __BUG_ON(test) __ASSERT_NO_MSG(test)
```
It would be good to add doxygen (and maybe a doxygen group) so that it appears in the right place on docs.zephyrproject.org .

Looks good to me though. Thanks for putting it together.

I'll hold off on approving until some others have reviewed. Maybe @peter-mitsis or @nashif?

xodus7 replied:
done

xodus7 committed:
test: kernel: fatal: exception: add test for __BUG

Adds some tests for the new __BUG macro with both asserts enabled
and disabled.

Signed-off-by: Corey Wharton <xodus7@cwharton.com>

xodus7 commented:
Added some ztest coverage.
